File: plugins/plugin-twitch/rust/src/providers/user_context.rs
//! User context provider for Twitch plugin.

use crate::types::{get_twitch_user_display_name, TwitchUserInfo};
use serde::{Deserialize, Serialize};

/// User context data
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserContextData {
    pub user_id: String,
    pub username: String,
    pub display_name: String,
    pub is_broadcaster: bool,
    pub is_moderator: bool,
    pub is_vip: bool,
    pub is_subscriber: bool,
    pub roles: Vec<String>,
    pub color: Option<String>,
}

/// User context values (simple key-value)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserContextValues {
    pub user_id: String,
    pub username: String,
    pub display_name: String,
    pub role_text: String,
    pub is_broadcaster: bool,
    pub is_moderator: bool,
}

/// User context provider result
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserContextResult {
    pub data: UserContextData,
    pub values: UserContextValues,
    pub text: String,
}

/// Get the current Twitch user context
pub fn get_user_context(user: &TwitchUserInfo, agent_name: &str) -> UserContextResult {
    let display_name = get_twitch_user_display_name(user);

    let mut roles = Vec::new();
    if user.is_broadcaster {
        roles.push("broadcaster".to_string());
    }
    if user.is_moderator {
        roles.push("moderator".to_string());
    }
    if user.is_vip {
        roles.push("VIP".to_string());
    }
    if user.is_subscriber {
        roles.push("subscriber".to_string());
    }

    let role_text = if roles.is_empty() {
        "viewer".to_string()
    } else {
        roles.join(", ")
    };

    let mut text = format!(
        "{} is talking to {} ({}) in Twitch chat.",
        agent_name, display_name, role_text
    );

    if user.is_broadcaster {
        text.push_str(&format!(" {} is the channel owner/broadcaster.", display_name));
    } else if user.is_moderator {
        text.push_str(&format!(" {} is a channel moderator.", display_name));
    }

    UserContextResult {
        data: UserContextData {
            user_id: user.user_id.clone(),
            username: user.username.clone(),
            display_name: display_name.clone(),
            is_broadcaster: user.is_broadcaster,
            is_moderator: user.is_moderator,
            is_vip: user.is_vip,
            is_subscriber: user.is_subscriber,
            roles,
            color: user.color.clone(),
        },
        values: UserContextValues {
            user_id: user.user_id.clone(),
            username: user.username.clone(),
            display_name,
            role_text,
            is_broadcaster: user.is_broadcaster,
            is_moderator: user.is_moderator,
        },
        text,
    }
}

/// Provider metadata
pub const PROVIDER_NAME: &str = "twitchUserContext";
pub const PROVIDER_DESCRIPTION: &str =
    "Provides information about the Twitch user in the current conversation";

File: plugins/plugin-twitch/src/providers/userContext.ts
/**
 * User context provider for Twitch plugin.
 */

import type {
  IAgentRuntime,
  Memory,
  Provider,
  ProviderResult,
  State,
} from "@elizaos/core";
import type { TwitchService } from "../service.js";
import {
  getTwitchUserDisplayName,
  TWITCH_SERVICE_NAME,
  type TwitchUserInfo,
} from "../types.js";

const MAX_TWITCH_FIELD_LENGTH = 200;

function truncateField(value: string | undefined): string | undefined {
  return value ? value.slice(0, MAX_TWITCH_FIELD_LENGTH) : value;
}
/**
 * Provider that gives the agent information about the Twitch user context.
 */
export const userContextProvider: Provider = {
  name: "twitchUserContext",
  description:
    "Provides information about the Twitch user in the current conversation",
  descriptionCompressed: "provide information Twitch user current conversation",

  dynamic: true,
  contextGate: { anyOf: ["social", "connectors"] },
  cacheStable: false,
  cacheScope: "turn",
  contexts: ["social", "connectors"],
  get: async (
    runtime: IAgentRuntime,
    message: Memory,
    state: State,
  ): Promise<ProviderResult> => {
    // Only provide context for Twitch messages
    if (message.content.source !== "twitch") {
      return {
        data: {},
        values: {},
        text: "",
      };
    }

    const twitchService =
      runtime.getService<TwitchService>(TWITCH_SERVICE_NAME);

    if (!twitchService?.isConnected()) {
      return {
        data: {},
        values: {},
        text: "",
      };
    }

    const agentName = state?.agentName || "The agent";

    // Try to get user info from message metadata
    const metadata = message.content.metadata as
      | Record<string, unknown>
      | undefined;
    const userInfo = metadata?.user as TwitchUserInfo | undefined;

    if (!userInfo) {
      return {
        data: {},
        values: {},
        text: "",
      };
    }

    try {
      const displayName =
        truncateField(getTwitchUserDisplayName(userInfo)) ?? "";
      const username = truncateField(userInfo.username) ?? userInfo.username;
      const roles: string[] = [];

      if (userInfo.isBroadcaster) {
        roles.push("broadcaster");
      }
      if (userInfo.isModerator) {
        roles.push("moderator");
      }
      if (userInfo.isVip) {
        roles.push("VIP");
      }
      if (userInfo.isSubscriber) {
        roles.push("subscriber");
      }

      const roleText = roles.length > 0 ? roles.join(", ") : "viewer";

      let responseText = `${agentName} is talking to ${displayName} (${roleText}) in Twitch chat.`;

      if (userInfo.isBroadcaster) {
        responseText += ` ${displayName} is the channel owner/broadcaster.`;
      } else if (userInfo.isModerator) {
        responseText += ` ${displayName} is a channel moderator.`;
      }

      return {
        data: {
          userId: userInfo.userId,
          username,
          displayName,
          isBroadcaster: userInfo.isBroadcaster,
          isModerator: userInfo.isModerator,
          isVip: userInfo.isVip,
          isSubscriber: userInfo.isSubscriber,
          roles,
          color: userInfo.color,
        },
        values: {
          userId: userInfo.userId,
          username,
          displayName,
          roleText,
          isBroadcaster: userInfo.isBroadcaster,
          isModerator: userInfo.isModerator,
        },
        text: responseText,
      };
    } catch (error) {
      return {
        data: {
          error: error instanceof Error ? error.message : String(error),
        },
        values: {},
        text: "",
      };
    }
  },
};
